File: dotnet-elasticemail-examples/Examples/Domains.cs
namespace ElasticEmailExamples;

using ElasticEmail.Api;
using ElasticEmail.Client;
using ElasticEmail.Model;
using Newtonsoft.Json;

public static class Domains
{
    public static async Task RunAsync()
    {
        var domainsApi = new DomainsApi(Ee.Config());
        var domain = Ee.SendingDomain;

        // 1. Add the domain
        try
        {
            await domainsApi.DomainsPostAsync(new DomainPayload(domain: domain));
            Console.WriteLine($"Domain \"{domain}\" added.");
        }
        catch (ApiException e) when (e.ErrorCode == 400
            && (Ee.ErrorMessage(e).Contains("exist", StringComparison.OrdinalIgnoreCase)
                || Ee.ErrorMessage(e).Contains("already", StringComparison.OrdinalIgnoreCase)))
        {
            Console.WriteLine($"Domain \"{domain}\" already exists.");
        }
        catch (ApiException e)
        {
            Ee.Fail("add domain", e);
        }

        // 2. Show DNS verification state. Add the records shown in the dashboard, then re-run.
        try
        {
            var data = await domainsApi.DomainsByDomainGetAsync(domain);
            Console.WriteLine("\nVerification status:");
            Console.WriteLine($"  SPF:       {(data.Spf ? "ok" : "missing")}");
            Console.WriteLine($"  DKIM:      {(data.Dkim ? "ok" : "missing")}");
            Console.WriteLine($"  MX:        {(data.MX ? "ok" : "missing")}");
            Console.WriteLine($"  DMARC:     {(data.DMARC ? "ok" : "missing")}");
            Console.WriteLine($"  Tracking:  {data.TrackingStatus?.ToString() ?? "n/a"}");
            Console.WriteLine($"  Default:   {(data.DefaultDomain ? "yes" : "no")}");
            if (data.DKIMRecord != null)
            {
                Console.WriteLine($"\nDKIM record to publish: {JsonConvert.SerializeObject(data.DKIMRecord)}");
            }
        }
        catch (ApiException e)
        {
            Ee.Fail("get domain", e);
        }

        // 3. List all domains
        try
        {
            var domains = await domainsApi.DomainsGetAsync();
            Console.WriteLine($"\nDomains on the account ({domains.Count}):");
            foreach (var d in domains)
            {
                Console.WriteLine($" - {d.Domain} spf={d.Spf} dkim={d.Dkim} default={d.DefaultDomain}");
            }
        }
        catch (ApiException e)
        {
            Ee.Fail("list domains", e);
        }

        // 4. Optional: verify tracking (CNAME must point at the Elastic Email tracking host)
        // await domainsApi.DomainsByDomainVerificationPutAsync(domain, "Http");

        // 5. Optional: set the default sender for the account
        // await domainsApi.DomainsByEmailDefaultPatchAsync($"hello@{domain}");

        Console.WriteLine("\nDone.");
    }
}

File: rust-elasticemail-examples/examples/domains.rs
use rust_elasticemail_examples::{already_exists, config, env_or, print_api_error};
use ElasticEmail::apis::domains_api;
use ElasticEmail::models::DomainPayload;

fn flag(value: Option<bool>) -> &'static str {
    if value.unwrap_or(false) { "ok" } else { "missing" }
}

#[tokio::main]
async fn main() {
    let config = config();
    let domain = env_or("SENDING_DOMAIN", "yourdomain.com");

    // 1. Add the domain
    let payload = DomainPayload { domain: Some(domain.clone()), ..DomainPayload::new() };
    match domains_api::domains_post(&config, payload).await {
        Ok(_) => println!("Domain \"{}\" added.", domain),
        Err(e) if already_exists(&e) => println!("Domain \"{}\" already exists.", domain),
        Err(e) => {
            print_api_error("add domain", &e);
            std::process::exit(1);
        }
    }

    // 2. Show DNS verification state. Add the records shown in the dashboard, then re-run.
    match domains_api::domains_by_domain_get(&config, &domain).await {
        Ok(d) => {
            println!("\nVerification status:");
            println!("  SPF:       {}", flag(d.spf));
            println!("  DKIM:      {}", flag(d.dkim));
            println!("  MX:        {}", flag(d.mx));
            println!("  DMARC:     {}", flag(d.dmarc));
            println!("  Tracking:  {}", d.tracking_status.map(|s| format!("{:?}", s)).unwrap_or_else(|| "n/a".into()));
            println!("  Default:   {}", if d.default_domain.unwrap_or(false) { "yes" } else { "no" });
            if let Some(record) = d.dkim_record {
                println!("\nDKIM record to publish: {}", serde_json::to_string(&record).unwrap_or_default());
            }
        }
        Err(e) => {
            print_api_error("get domain", &e);
            std::process::exit(1);
        }
    }

    // 3. List all domains
    match domains_api::domains_get(&config).await {
        Ok(domains) => {
            println!("\nDomains on the account ({}):", domains.len());
            for d in domains {
                println!(
                    " - {} spf={} dkim={} default={}",
                    d.domain.unwrap_or_default(),
                    d.spf.unwrap_or(false),
                    d.dkim.unwrap_or(false),
                    d.default_domain.unwrap_or(false)
                );
            }
        }
        Err(e) => {
            print_api_error("list domains", &e);
            std::process::exit(1);
        }
    }

    // 4. Optional: verify tracking (CNAME must point at the Elastic Email tracking host)
    // domains_api::domains_by_domain_verification_put(&config, &domain, "Http").await.unwrap();

    // 5. Optional: set the default sender for the account
    // domains_api::domains_by_email_default_patch(&config, &format!("hello@{}", domain)).await.unwrap();

    println!("\nDone.");
}
